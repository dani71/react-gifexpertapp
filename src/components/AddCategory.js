import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ({ setCategories }) => {

    //const [inputValue, setInputValue] = useState();// esto da error, porque es undefined
    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (e) => {
        // console.log(e.target.value);
        setInputValue(e.target.value);
        // console.log('HandelInputChange llamado');
    }

    const handleSubmit = (e) => {  
        e.preventDefault();
        if (inputValue.trim().length > 2) { //evita la categoría vacía
            // console.log(setCategories)
            //setCategories((c) => {[...c, e]});
            setCategories((cats) => [inputValue, ...cats]);
            setInputValue(''); // evita el escribir dos veces lo mismo initencionadamente
        };

    }
    return (
        <form onSubmit={handleSubmit}>
            <input
                type="text"
                value={inputValue}
                onChange={handleInputChange}
            />
        </form>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}