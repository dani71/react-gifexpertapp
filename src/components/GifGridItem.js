import React from 'react'

export const GifGridItem = ( {id, title, url} ) => {
 
    console.log({id, title, url})
    return (
        <div className="card animate__animated animate__fadeInDown"> {/*class puede ser para html o para JS, as√≠ que la han cambiado a className para estar seguros*/}
            <img src={url} alt={title}/>
            <p>{title}</p>
        </div>
    )
}
// class="animate__animated animate__bounce animate__delay-2s"
