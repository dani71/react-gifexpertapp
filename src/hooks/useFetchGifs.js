import { useEffect, useState } from "react"
import { getGifs } from "../helpers/getGifs";

export const useFetchGifs = ( category ) => {

    const [state, setState] = useState( {
        data: [],
        loading: true
    });

    useEffect(() => { //los efectos no pueden ser async, pero la respuesta getGifs si es una promesa

        getGifs( category )
            .then( imgs => {
                setState( {
                    data: imgs,
                    loading: false
                });
            })
    }, [ category ])
        // useEffect(() => {
    //     getGifs( category )
    //     .then( setImages ); // esto es lo mismo que .then( imgs => setImages( imgs ))
    // }, [ category ]) //aquÃ­ se pasa una lista de dependencias, y si no tiene no lanza otras peticiones 
    //                     // si la categorÃ­a cambia se dispara otra vez.  

    // setTimeout( () => {
    //     setState ( {
    //         data: [1,2,3,4,5,6,7],
    //         loading : false
    //     })
    // }, 3000)
    return state; // {data:[], loading:true};
}

