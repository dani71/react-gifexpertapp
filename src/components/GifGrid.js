import React from 'react'
import { useFetchGifs } from '../hooks/useFetchGifs'
// import React, { useState, useEffect } from 'react'
// import { getGifs } from '../helpers/getGifs'
import { GifGridItem } from './GifGridItem'


export const GifGrid = ({ category }) => {

    // const [images, setImages] = useState([])

    // useEffect(() => {
    //     getGifs( category )
    //     .then( setImages ); // esto es lo mismo que .then( imgs => setImages( imgs ))
    // }, [ category ]) //aquí se pasa una lista de dependencias, y si no tiene no lanza otras peticiones 
    //                     // si la categoría cambia se dispara otra vez.    

    const { data: images, loading } = useFetchGifs(category);

    // console.log(loading);
    // console.log(data);

    return (
        <>
            <h3 className="animate__animated animate__fadeInDown">{category}</h3>
            {loading && <p className='animate__animated animate__flash'>Loading</p> /*solo hace algo si está cargando*/}
            <div className='card-grid'>
                {
                    images.map(img => (
                        <GifGridItem
                            key={img.id}
                            {...img}
                        />
                    ))
                }
            </div>
        </>
    )
}
