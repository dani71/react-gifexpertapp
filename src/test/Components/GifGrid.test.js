import React from 'react';
import { shallow } from "enzyme";
import '@testing-library/jest-dom'; //para tener la ayuda en el expect
import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');


describe('Pruebas sobre <GifGrid />', () => {

    const category = 'One Punch'

    test('Debe ser igual al snapshot ', () => {

        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });

        const wrapper = shallow(<GifGrid category={category} />);
        expect(wrapper).toMatchSnapshot();

    });

    test('debe de mostrar items cuando se cargan imágenes con useFecthGifs', () => { //deben ser pruebas unitarias, sin mezclarse con otros elementos => con un mock

        const gifs = [{
            id: 'ABC',
            url: 'https://localhost/cualquier/cosa.jpg',
            title: ' cualquier cosa'
        }];

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        const wrapper = shallow(<GifGrid category={category} />);
        expect(wrapper).toMatchSnapshot();
    });

    // si está en false no debería existir el párrafo    
    test('si loading está en false no debería existir el párrafo ', () => {

        const gifs = [{
            id: 'ABC',
            url: 'https://localhost/cualquier/cosa.jpg',
            title: ' cualquier cosa'
        }, {
            id: 'ABCsasdfas',
            url: 'https://localhost/cualquier/cosa.jpg',
            title: ' cualquier cosa'
        }];

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        const wrapper = shallow(<GifGrid category={category} />);
        expect(wrapper.find('p').exists()).toBe(false);
        expect(wrapper.find('GifGridItem').length).toBe(gifs.length);
    })
})
