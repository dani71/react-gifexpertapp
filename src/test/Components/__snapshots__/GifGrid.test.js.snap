// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre <GifGrid /> Debe ser igual al snapshot  1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeInDown"
  >
    One Punch
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`Pruebas sobre <GifGrid /> debe de mostrar items cuando se cargan im√°genes con useFecthGifs 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeInDown"
  >
    One Punch
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title=" cualquier cosa"
      url="https://localhost/cualquier/cosa.jpg"
    />
  </div>
</Fragment>
`;
